### PR TITLE
SPR-14664 Add byte range requests support

### PR DESCRIPTION
**Pull request by bclozel**
Add byte range requests support for static resources in the reactive
web project.

### CODE REVIEW

File: spring-core/src/main/java/org/springframework/util/ResourceUtils.java [hunk 1]
```diff
@@ -384,4 +389,23 @@ public static void useCachesIfNecessary(URLConnection con) {
...
+		if (InputStreamResource.class != resource.getClass()) {
+			try {
+				return OptionalLong.of(resource.contentLength());
+			}
+			catch (IOException ignored) {
+			}
+		}
+		return OptionalLong.empty();
+	}
+
```
Comment (rstoyanchev) [inline spring-core/src/main/java/org/springframework/util/ResourceUtils.java]: Could we consider a new method on Resource for this? The fact that a utility method takes a Resource and calls its property is a sign. It looks like a variant of the current `getContentLength()` that returns the content length optionally if known. So perhaps `OptoinalLong getContentLengthIfKnown()` with a default implementation returning empty.

/cc @jhoeller, @poutsma

**Comment by rstoyanchev** [comment]
Only one minor comment about on ResourceUtils. Aside from that looks very good and once the `ServerHttpMessageWriter` is in place it will be even more encapsulated and re-usable from both the `ResourceWebHandler` and for `ResponseBody` methods.

File: spring-core/src/main/java/org/springframework/util/ResourceUtils.java (hunk 1 above)
Comment (jhoeller) [inline spring-core/src/main/java/org/springframework/util/ResourceUtils.java]: I don't mind an addition to `Resource`, in particular now where we can do default methods. However, I feel that we're overusing `Optional`. On such core abstractions, can we rather do it in a more traditional style through a `Long` return type which may simple be `null` if not known?

To be clear, using `Optional` in a functional API is perfectly sensible. I just don't want it to creep into Java interfaces at a lower level, in particular not in existing interfaces with a traditional design mindset in other methods... and not really in new interfaces in the same packages either.

Comment (rstoyanchev) [inline spring-core/src/main/java/org/springframework/util/ResourceUtils.java]: For consistency with the existing method, possibly returning `long` with -1 to indicate unknown. It looks like some (e.g. UrlConnection) can return -1.

**Comment by bclozel** [comment]
I just tried reusing the `ServerHttpMessageWriter` and it seems we can't really move the `HttpRange` information from `ResourceWebHandler` because:

* In `ResourceWebHandler`, in case an `IllegalArgumentException` is thrown while parsing the HTTP range  information, we not only need to set an HTTP status (which could be possible with a `ResponseStatusException`), but also add an HTTP response header.
* In `ResourceWebHandler`, we're setting the response status to `HttpStatus.PARTIAL_CONTENT` if this is a byte range request; we can't do that from within an `HttpMessageWriter` which only holds a reference to a `ReactiveHttpOutputMessage` (no `setStatus` method).

Given that `ResourceHttpMessageWriter` already delegates to `ResourceRegionHttpMessageWriter`, which delegates to the `ResourceRegionEncoder`, I'm wondering if adding another level of wrapping and hints is really helping the situation in the end.

**Comment by rstoyanchev** [comment]
@sdeleuze in light of @bclozel's comment I'd like to propose something I was already thinking about. Instead of `ServerHttpMessageWriter` having a method to prepare hints, it should have a method to write except taking all the inputs that are available on the server side (e.g. the type of the stream, the ServerHttpRequest and ServerHttpResponse). It can then prepare hints and delegate to the next writer, or compose on the input/output stream, set the status, short-circuit the process, and so on.

**Comment by bclozel** [comment]
@rstoyanchev @sdeleuze to me, this `ServerHttpMessageWriter` contract would look more and more like a full `WebHandler` with a few things added on top. This makes sense for this use case, but I'm not sure I'm comfortable adding this right now (I've explained in my previous comment the level of indirection we're talking about if we do that).

I'd rather duplicate a few lines of code in `ResourceWebHandler` and a `HandlerResultHandler` implementation rather than introducing that abstraction if we're not confident about it. Will it be useful for more use cases?

**Comment by rstoyanchev** [comment]
Well I'm not suggesting it because of the Resource use case alone. I was independently thinking of making this suggestion. Plus we are not talking about a major change to what's already in place that we are comfortable with. I'm simply suggesting to replace `resolveHints` with a `write`, so pretty much the same story but a little more control. 

If we accept that an `HttpMessageWriter` is intentionally limited to what is relevant to both client and server and that a `ServerHttpMessageWriter` is what `HttpMessageWrite` could be if we could pass to it the input available on the server side, then the contracts should be largely the same with a few extra inputs. In short I think resolvHints constrains things a little more than necessary, so you can only produce hints given server-side inputs, and I don't see a strong reason for that.

**Comment by sdeleuze** [comment]
@rstoyanchev Is your proposal to replace `ServerHttpMessageWriter#resolveWriteHints()` by something like
`Mono<Void> ServerHttpMessageWriter#write(Publisher<? extends T> inputStream, ResolvableType streamType, ResolvableType elementType, MediaType mediaType, ServerHttpRequest request, ServerHttpResponse response, Map<String, Object> hints)` where `hints` could be used to read existing hints as well as adding eventually new ones before delegating to the nested writer?

**Comment by rstoyanchev** [comment]
@sdeleuze yes that's what I was thinking about.

**Comment by sdeleuze** [comment]
Here is a draft just on the writer to discuss: https://github.com/sdeleuze/spring-framework/commit/b17f596f60fac7af458e6cf5b3cfcb911d0c3744

Main points are:
 - `hints` are not provided anymore to `canDecode()` / `canWrite()` because we don't have a dedicated method to get them anymore (not an issue for my POV, hint should just be used to customize the output)
 - `ServerHttpMessageWriter` now has a more powerful write method
 - For `AbstractServerHttpMessageWriter` I made various tries because letting implementers directly implement `write()` or `writeInternal()` lead to not so obvious and error prone implementations and I was not happy with that. I ended-up with this `beforeWrite()` solution that allows easily to provide hints or set the status. For really advanced use cases, it is still possible to override `write()` if needed.

Any thoughts ?

**Comment by rstoyanchev** [comment]
Looks good to me, even a bit simpler with hints now prepared and managed internally.

**Comment by sdeleuze** [comment]
These improvements are now in master.

@bclozel Let me know if there is any issue while rebasing your changes.

**Pull request merged**